prow/gerrit/adapter/adapter.go

```diff
@@ -0,0 +1,215 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package adapter implements a controller that interacts with gerrit instances
+package adapter
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/gerrit/client"
+	"k8s.io/test-infra/prow/kube"
+	"k8s.io/test-infra/prow/pjutil"
+)
+
+type kubeClient interface {
+	CreateProwJob(kube.ProwJob) (kube.ProwJob, error)
+}
+
+type gerritClient interface {
+	QueryChanges(lastUpdate time.Time, rateLimit int) map[string][]gerrit.ChangeInfo
+	SetReview(instance, id, revision, message string) error
+}
+
+type configAgent interface {
+	Config() *config.Config
+}
+
+// Controller manages gerrit changes.
+type Controller struct {
+	ca configAgent
+	kc kubeClient
+	gc gerritClient
+
+	storage string
+
+	lastUpdate time.Time
+}
+
+// NewController returns a new gerrit controller client
+func NewController(storage string, projects map[string][]string, kc *kube.Client, ca *config.Agent) (*Controller, error) {
+	lastUpdate := time.Now()
+	if storage != "" {
+		buf, err := ioutil.ReadFile(storage)
+		if err == nil {
+			unix, err := strconv.ParseInt(string(buf), 10, 64)
+			if err != nil {
+				return nil, err
+			}
+			lastUpdate = time.Unix(unix, 0)
+		} else if !os.IsNotExist(err) {
+			return nil, err
+		}
+		// fallback to time.Now() if file does not exist yet
+	}
+
+	c, err := gerrit.NewClient(projects)
+	if err != nil {
+		return nil, err
+	}
+	c.Start()
+
+	return &Controller{
+		kc:         kc,
+		ca:         ca,
+		gc:         c,
+		lastUpdate: lastUpdate,
+		storage:    storage,
+	}, nil
+}
+
+// SaveLastSync saves last sync time in Unix to a volume
```

Comment:
hmm why? this sounds like a bad idea?

Reply:
discussed offline this is the _gerrit_ sync, which we need because there are no webhooks so we need to figure out where to scan events from >.>